kbchat/kbchat_test.go

```diff
@@ -103,82 +102,284 @@ func deleteWorkingDir(workingDir string) error {
 	return os.RemoveAll(workingDir)
 }
 
-var kbc *API
-var config Config
+type testSetupOptions struct {
+	OneOnOnePartner string
+	TeamName        string
+}
 
-func TestMain(m *testing.M) {
-	var err error
-	config, err = readAndParseConfig()
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error in reading config: %v\n", err)
-	}
-	dir, err := randomTempDir()
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error in generating directory: %v\n", err)
+func testSetup(botName string, options *testSetupOptions) (bot *API, config botConfig, dir string, oneOnOneChannel Channel, teamChannel Channel) {
+	var oneOnOnePartner string
+	var teamName string
+	if options == nil {
+		oneOnOnePartner = "charlie1"
+		teamName = "acme"
+	} else {
+		oneOnOnePartner = options.OneOnOnePartner
+		teamName = options.TeamName
 	}
+
+	config = readAndParseConfig()
+	dir = randomTempDir()
 	kbLocation, err := prepWorkingDir(dir)
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error in preparing working directory: %v\n", err)
+		defer deleteWorkingDir(dir)
+		panic(err)
 	}
-	kbc, err = Start(RunOptions{KeybaseLocation: kbLocation, HomeDir: dir, Oneshot: config.Bots.Alice1, StartService: true})
+	bot, err = Start(RunOptions{KeybaseLocation: kbLocation, HomeDir: dir, Oneshot: config.Bots[botName], StartService: true})
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Error in starting service: %v\n", err)
+		defer testTeardown(bot, dir)
+		panic(err)
 	}
 
-	flag.Parse()
-	code := m.Run()
+	oneOnOneChannel = Channel{
+		Name: fmt.Sprintf("%s,%s", config.Bots[botName].Username, oneOnOnePartner),
+	}
+	teamChannel = Channel{
+		Name:        config.Teams[teamName].Name,
+		Public:      false,
+		MembersType: "team",
+		TopicName:   config.Teams[teamName].TopicName,
+		TopicType:   "chat",
+	}
+
+	return bot, config, dir, oneOnOneChannel, teamChannel
```

Comment:
this is a lot of stuff being returned. do we _need_ all of this? if we split into two, we only ever need to return one of these channels. the `dir` looks necessary for cleanup. i think we're stuck with it. i don't love returning `config`. looks like it's only being used for the username of the bot. huh. it's an unfortunate leaking of the underlying information just for that one thing. maybe it makes sense for `TestGetUsername` also to call into whatever is parsing the yaml config.